File: kattis/Exercise2/ChoppingWood.cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<int> v(n);
    // Frequencies for vertices 1..n+1 (we use size n+2 for safe indexing)
    vector<int> freq(n + 2, 0);

    // Read the 'v' column and check range
    for(int i = 0; i < n; i++){
        cin >> v[i];
        if(v[i] < 1 || v[i] > n + 1){
            cout << "Error\n";
            return 0;
        }
        freq[v[i]]++;
    }

    // A min-structure of leaves (vertices with freq == 0)
    set<int> leaves;
    for(int x = 1; x <= n + 1; x++){
        if(freq[x] == 0){
            leaves.insert(x);
        }
    }

    vector<int> ucol(n); // to store the reconstructed u-column

    // Simulate the leaf-picking process
    for(int i = 0; i < n; i++){
        if(leaves.empty()){
            // no leaf available => inconsistent
            cout << "Error\n";
            return 0;
        }
        // pick the smallest leaf
        int u = *leaves.begin();
        leaves.erase(leaves.begin());
        ucol[i] = u;

        // decrement freq of v[i]
        int w = v[i];
        freq[w]--;
        if(freq[w] < 0){
            // invalid, used w too many times
            cout << "Error\n";
            return 0;
        }
        // if w is now a leaf, add it
        if(freq[w] == 0){
            leaves.insert(w);
        }
    }

    // after n removals, exactly one vertex should remain
    if(leaves.size() != 1){
        cout << "Error\n";
        return 0;
    }

    // output the reconstructed u-column
    for(int i = 0; i < n; i++){
        cout << ucol[i] << "\n";
    }
    return 0;
}

File: kattis/Exercise2/ChoppingWood.py
def solve():
    n = int(input())
    V = [int(input()) for _ in range(n)]

    freq = [0] * (n + 2)
    for v in V:
        if v < 1 or v > n + 1:
            print("Error")
            return
        freq[v] += 1


    heap = []
    heap_size = 0

    def heap_push(x):
        nonlocal heap_size
        heap.append(x)
        heap_size += 1
        i = heap_size - 1
        while i > 0:
            p = (i - 1) // 2
            if heap[p] > heap[i]:
                heap[p], heap[i] = heap[i], heap[p]
                i = p
            else:
                break

    def heap_pop():
        nonlocal heap_size
        if heap_size == 0:
            return None
        ret = heap[0]
        heap_size -= 1
        if heap_size > 0:
            heap[0] = heap[heap_size]
            i = 0
            while True:
                left = 2 * i + 1
                right = 2 * i + 2
                if left >= heap_size:
                    break
                smallest = left
                if right < heap_size and heap[right] < heap[left]:
                    smallest = right
                if heap[smallest] < heap[i]:
                    heap[i], heap[smallest] = heap[smallest], heap[i]
                    i = smallest
                else:
                    break
        heap.pop()  
        return ret

    for i in range(1, n + 1):
        if freq[i] == 0:
            heap_push(i)

    U = []  
    for v in V:
        u = heap_pop()
        if u is None:
            print("Error")
            return
        U.append(u)
        if freq[v] < 1:
            print("Error")
            return
        freq[v] -= 1
        if v <= n and freq[v] == 0:
            heap_push(v)

    for x in freq:
        if x != 0:
            print("Error")
            return

    for leaf in U:
        print(leaf)
solve()